File: x402-payment-agent/python/vendors_store.py
"""
vendors.csv — allowlist of vendors approved for autonomous payment.

Schema:
  vendor_name, vendor_email, max_amount_usd, notes

A vendor is "trusted" if their email (case-insensitive) appears in this CSV.
Their per-vendor `max_amount_usd` overrides the global cap (lower of the two
applies). Re-read on every iteration so live edits to the file take effect.
"""

import csv
from pathlib import Path

FILE = Path("vendors.csv")


def load() -> list[dict]:
    if not FILE.exists():
        return []
    with FILE.open("r", newline="") as f:
        return [
            {
                "vendor_name": (r.get("vendor_name") or "").strip(),
                "vendor_email": (r.get("vendor_email") or "").strip().lower(),
                "max_amount_usd": float(r.get("max_amount_usd") or 0),
                "notes": (r.get("notes") or "").strip(),
            }
            for r in csv.DictReader(f)
        ]


def find(vendors: list[dict], sender_email: str) -> dict | None:
    """Look up a vendor by their email."""
    if not sender_email:
        return None
    target = sender_email.lower().strip()
    for v in vendors:
        if v["vendor_email"] == target:
            return v
    return None

File: typescript/src/vendorsStore.ts
/**
 * vendors.csv — allowlist of vendors approved for autonomous payment.
 */

import { existsSync, readFileSync } from "node:fs";

const FILE = "vendors.csv";

export interface Vendor {
  vendor_name: string;
  vendor_email: string;
  max_amount_usd: number;
  notes: string;
}

function parseLine(line: string): string[] {
  const out: string[] = [];
  let cur = "", inQuote = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (inQuote) {
      if (c === '"' && line[i + 1] === '"') { cur += '"'; i++; }
      else if (c === '"') inQuote = false;
      else cur += c;
    } else {
      if (c === '"') inQuote = true;
      else if (c === ",") { out.push(cur); cur = ""; }
      else cur += c;
    }
  }
  out.push(cur);
  return out;
}

export function load(): Vendor[] {
  if (!existsSync(FILE)) return [];
  const lines = readFileSync(FILE, "utf-8").split("\n").filter(l => l.length > 0);
  if (lines.length < 2) return [];
  const headers = parseLine(lines[0]);
  return lines.slice(1).map(line => {
    const fields = parseLine(line);
    const row: any = {};
    headers.forEach((h, i) => row[h] = fields[i] ?? "");
    return {
      vendor_name: (row.vendor_name || "").trim(),
      vendor_email: (row.vendor_email || "").trim().toLowerCase(),
      max_amount_usd: parseFloat(row.max_amount_usd) || 0,
      notes: (row.notes || "").trim(),
    } as Vendor;
  });
}

export function find(vendors: Vendor[], senderEmail: string): Vendor | null {
  if (!senderEmail) return null;
  const target = senderEmail.toLowerCase().trim();
  return vendors.find(v => v.vendor_email === target) ?? null;
}
